Support for button and axis names

leon-gh opened the pull request:
* Created an external device_db.json file which contains mapping of buttons to their names, e.g. F-16 stick might have a Button 10 which is known as TMS UP, so Gremlin would be able to display button info as Button 10 - TMS UP
* Added configurable option to display button names as:
  - None - Button ID
  - Mixed - Button ID + Button name
  - Name - Button name

leon-gh committed:
Support for button and axis names

* Created an external device_db.json file which contains mapping
  of buttons to their names, e.g. F-16 stick might have a
  Button 10 which is known as TMS UP, so Gremlin would
  be able to display button info as Button 10 - TMS UP
* Added configurable option to display button names as:
  - None - Button ID
  - Mixed - Button ID + Button name
  - Name - Button name

WhiteMagic commented on gremlin/ui/device.py:
```diff
@@ -30,13 +31,108 @@
+
+    def _load(self) -> None:
+
+        db_file = "device_db.json"
```
Probably will need to use `util.resource_path()` here such that the file can be found when the whole thing is "compiled" with PyInstaller.

leon-gh replied:
n/p

WhiteMagic commented on gremlin/ui/device.py:
```diff
@@ -30,13 +31,108 @@
+            except ValueError:
+                pass
+
+        if not load_successful:
```
We might want to do a check on the contents to catch obvious mistakes like misspelt keys. So `button 1` doesn't work but `Button 1` does which we can easily catch, and the same if there is a mapping that is missing one end we can also emit warnings to the logfile.

This could also be done via Json Schema but I feel that this might be overkill :-)

leon-gh replied:
ok

WhiteMagic commented on gremlin/ui/device.py:
```diff
@@ -30,13 +31,108 @@
+        input name, e.g. Axis 1 or Button 1.
+
+        Some devices have configurable number of buttons and/or axes so
+        adjusting device database information might be required in the future.
```
I usually add argument and return value annotation as well in the following manner:
```
Args:
    input_name: ...
    input_map: ...

Returns:
    ...
```
This will allow sphinx to later on generate useful documentation of everything, something R13 was missing. And thankfully we don't need the weird type annotation since typing now contains that info.

leon-gh replied:
n/p

WhiteMagic commented on gremlin/ui/device.py:
```diff
@@ -30,13 +31,108 @@
+        if input_name_display_mode == "None" or input_name not in input_map:
+            return input_name
+
+        db_input_name = input_map[input_name].strip()
```
This will fail if the mapping is incomplete, i.e. `input_name` does not exist in the mapping. Using `.get(input_name, "")` should work around that issue.

leon-gh replied:
Line 93 makes sure that input_name is in the mapping. If you want me to refactor it into .get, no issues, but line 93 checks whether input name is in the input_map and returns a default, if not.

WhiteMagic replied:
Ahh yeah I think I realized that the first time around and the second time I didn't see it. I think that's fine, the only thing I could think of to "disarm" that confusion is to have all that follows after in the `else` branch of the check in 93 but not sure if that makes it worse.

WhiteMagic commented on gremlin/ui/device.py:
```diff
@@ -30,13 +31,108 @@
+                d["product"] == device.product_id and
+                d["vendor"] == device.vendor_id
+            ):
+                if (
```
Probably can be simplified to something like `d.get("mapping", None) in self.device_db["mapping|]` I would assume that `None` won't be found in the mapping dict.

leon-gh replied:
ok

WhiteMagic commented on gremlin/ui/device.py:
```diff
@@ -194,6 +290,7 @@ def __init__(self, parent=None):
         super().__init__(parent)
 
         self._device: dill.DeviceSummary | None = None
+        self.device_mapping: Dict[str, str] | None = None
```
Since we don't expect to expose this I'd turn it into a private variable, i.e. `self._device_mapping`

leon-gh replied:
ok

WhiteMagic commented on gremlin/ui/device.py:
```diff
@@ -997,4 +1100,18 @@ def axisIdentifier(self, index: int) -> int:
+    "Mixed",
+    "Defines how input name is displayed.",
+    {
+        "valid_options": ["None", "Mixed", "Name"]
```
"None" is a somewhat confusing label as it might be interpreted as showing no labels at all. Maybe "Numerical", "Numerical & Label", "Label" might be clearer?

leon-gh replied:
Yup, I agree.

WhiteMagic commented on gremlin/ui/device.py:
```diff
@@ -30,13 +31,108 @@
 
+@SingletonDecorator
+class DeviceDatabase:
+    """Provides button/axis to name mapping for known devices"""
```
I assume hats would be in here too as we don't restrict the code?

leon-gh replied:
Yes, Hats are supported too.

WhiteMagic commented on gremlin/ui/device.py:
```diff
@@ -30,13 +31,108 @@
+    def __init__(self) -> None:
+        self._load()
+
+    def _load(self) -> None:
```
Technically could be part of the constructor, though allows us to reload it in the future if needed.

leon-gh replied:
Originally I had it in the constructor, but I don't like long constructors and made it into a method. And yes, we could reload if contents changes in the future.

WhiteMagic replied:
Yeah I feel the same, so keeping it as is sounds good to me.

WhiteMagic commented on gremlin/ui/device.py:
```diff
@@ -30,13 +31,108 @@
+
+        load_successful = False
+        json_data = {}
+        with open(db_file) as hdl:
```
Is there any benefit to this over a more concise `json_data = json.load(open(...))` in a try except block?

leon-gh replied:
No, I usually used json.load(), but saw this method in gremlin/config.py and reused it. If json.load() is preferred, I can refactor it.

WhiteMagic replied:
That code probably was written eons ago and I wouldn't know why I decided to do it that way, so I think doing the json.load() bit is better as that's what I'd gravitate towards nowadays too.

leon-gh replied:
OK, will refactor.

WhiteMagic commented on gremlin/ui/device.py:
```diff
@@ -30,13 +31,108 @@
+                    "please fix the errors and restart the application.")
+
+        # create an empty device database if load was not successful
+        self.device_db = json_data if load_successful else {}
```
Does this not having "device" and "mapping" keys cause issues?

leon-gh replied:
Good catch. Added defaults.

WhiteMagic commented on gremlin/ui/device.py:
```diff
@@ -283,10 +381,15 @@ def inputIdentifier(self, index: int) -> InputIdentifier:
+            )
+
+        if self.device_mapping is not None:
+            return DeviceDatabase().input_name(input_name, self.device_mapping)
```
Since we get a mapping might be nicer for this to be an actual object we can just use rather than having to go through the database each time as it's a bit weird that we give that singleton it's own data back.

leon-gh replied:
Refactored into DeviceMapping object.

leon-gh committed:
Addressed review comments

* attributes made _private
* refactored as suggested into DeviceMapping object
* added simple sanity check on device_db.json structure
* renamed product/vendor in device_db to product_id/vendor_id
  to make them consistent with dill device information

leon-gh committed:
Review comments

* no code changes, updated descriptive text only

leon-gh committed:
Review comments

* refactored json loading into json.loads(open(...))

WhiteMagic commented on gremlin/ui/device.py:
```diff
@@ -30,13 +32,164 @@
+        if (
+            device_db.get("mapping", None) is None or
+            device_db.get("devices", None) is None or
+            not isinstance(device_db.get("mapping", []), dict) or
```
Is this intended to be a mismatched default return type for `get` on mapping? The devices one is [] -> list while mapping is [] -> dict.

WhiteMagic commented on gremlin/ui/device.py:
```diff
@@ -30,13 +32,164 @@
+        device_count = 0
+        for dev in device_db.get("devices", []):
+            if (
+                dev.get("product_id", None) is None or
```
Technically this could be made a bit more concise and possibly clearer with this to not have to rely on getting `None` via default value.
`if not("product_id" in dev and "vendor_id" in dev and "mapping" in dev):`

WhiteMagic commented on gremlin/ui/device.py:
```diff
@@ -30,13 +32,164 @@
+            else:
+                device_count += 1
+
+        mapping_count = 0
```
Would this be the same as `mapping_count = len(device_db.get("mapping", [])`?

WhiteMagic commented on gremlin/ui/device.py:
```diff
@@ -30,13 +32,164 @@
+                dev.get("vendor_id", None) == device.vendor_id
+            ):
+                if (
+                    "mapping" not in dev or
```
`"mapping" not in dev` should not occur anymore due to the `_parse_device_db` check, correct?

WhiteMagic commented:
Overall looks good to me, added a few small comments/clarification questions and after that this should be good in my eyes.

leon-gh committed:
Addressed review comments

* refactored the code for readability, removed unnecessary checking code which is no longer
  required due to the fact that _load() produces a valid JSON with "devices" and "mapping" objects.
